### PR TITLE
ChangeSignature: enable on local functions

### PR DESCRIPTION
**Pull request by jcouv**
### Customer scenario

1. Write a local function
2. Invoke the light bulb (Ctrl+.) while the cursor is on one its parameters
3. Expect ChangeSignature to come up

### Bugs this fixes
Fixes https://github.com/dotnet/roslyn/issues/14122
Fixes https://github.com/dotnet/roslyn/issues/8359

### Risk & Performance impact
Low. The ChangeSignature service can already handle many kinds of signatures (methods, delegates, constructors, invocations, etc) and this is just adding one more. The new logic tells the ChangeSignature service to recognize local function statements and directs it to the local function's parameter list.

### Is this a regression from a previous update?
No

### Root cause analysis
Local functions where introduced in C# 7.0, but with minimal IDE support.

### How was the bug found?
Reported internally.

image

image

Note: as shown in the second screenshot, there is a minor wart with the ChangeSignature popup: it should probably omit "private static" when operating on local functions. But I couldn't figure out whether that code is in Roslyn.

### CODE REVIEW

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 1]
```diff
@@ -258,6 +260,13 @@ private SyntaxNode GetNodeContainingTargetNode(SyntaxNode matchingNode)
                 return method.WithParameterList(method.ParameterList.WithParameters(updatedParameters).WithAdditionalAnnotations(changeSignatureFormattingAnnotation));
             }
```
Comment (Neme12) [inline src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: should local functions work with `UpdateParamTagsInLeadingTrivia` above?

File: src/EditorFeatures/CSharpTest/ChangeSignature/ReorderParametersTests.cs [hunk 2]
```diff
@@ -33,6 +33,70 @@ public void Goo(string y, int x)
             await TestChangeSignatureViaCommandAsync(LanguageNames.CSharp, markup, updatedSignature: permutation, expectedUpdatedInvocationDocumentCode: updatedCode);
         }
 
+        [WpfFact, Trait(Traits.Feature, Traits.Features.ChangeSignature)]
+        public async Task ReorderLocalFunctionParameters()
```
Comment (Neme12) [inline src/EditorFeatures/CSharpTest/ChangeSignature/ReorderParametersTests.cs]: would be nice to move this below ReorderMethodParametersAndArguments and call it ReorderLocalFunctionParametersAndArguments for consistency

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs (hunk 1 above)
Comment (jcouv) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: Good question. At this point, no. Local functions don't have xml docs.

Comment (Neme12) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: hmmm... i tried typing `///` and it created one for me. probably a bug then?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: Not really.  we're not going to stop you from writing the doc comment if you want.  After all, you may intend to move this to a class-method.  (or, it just may be you're missing a } and perhaps you meant this to be a class member).  I'm not sure we'd go out of our way to make this work.  but if it works, we'll keep allowing it to.

**Comment by CyrusNajmabadi** [comment]
> But I couldn't figure out whether that code is in Roslyn.

It most likely is.  I can't imagine any other place that would have any clue how to add that.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs (hunk 1 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: actually i'm not able to reproduce it anymore

**Comment by jcouv** [comment]
Found the code for dialog (by searching for some of the strings). It's in `ChangeSignatureDialog`. I'll try to improve the signature preview.

**Comment by jcouv** [comment]
Adjusted the symbol display and verified in ChangeSignature window:
image

**Comment by jcouv** [comment]
@agocke for compiler-side review (adjusting symbol display for local functions)

**Comment by jcouv** [comment]
@dotnet/roslyn-ide for review. Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-ide for review. Thanks

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs [hunk 3]
```diff
@@ -287,12 +288,22 @@ private void AddAccessibilityIfRequired(ISymbol symbol)
...
             if (format.MemberOptions.IncludesOption(SymbolDisplayMemberOptions.IncludeAccessibility) &&
                 (containingType == null ||
-                 (containingType.TypeKind != TypeKind.Interface && !IsEnumMember(symbol))))
+                 (containingType.TypeKind != TypeKind.Interface && !IsEnumMember(symbol) & !IsLocalFunction(symbol))))
             {
                 AddAccessibility(symbol);
             }
         }
 
+        private bool IsLocalFunction(ISymbol symbol)
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs]: :bulb: `static`

File: src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs [hunk 4]
```diff
@@ -5854,6 +5854,47 @@ void M()
...
+{
+    void M()
+    {
+        void Local() {}
+    }
+}");
+            var root = srcTree.GetRoot();
+            var comp = CreateCompilation(srcTree);
+
+            var semanticModel = comp.GetSemanticModel(comp.SyntaxTrees.Single());
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs]: `GetSemanticModel(srcTree)`

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 5]
```diff
@@ -258,6 +260,13 @@ private SyntaxNode GetNodeContainingTargetNode(SyntaxNode matchingNode)
...
 
+            if (updatedNode.IsKind(SyntaxKind.LocalFunctionStatement))
+            {
+                var localFunction = updatedNode as LocalFunctionStatementSyntax;
```
Comment (cston) [inline src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: Use cast rather than `as`.

**Comment by jcouv** [comment]
windows_release_unit32_prtest failed ([details](https://ci.dot.net/job/dotnet_roslyn/job/dev15.7.x/job/windows_release_unit32_prtest/573/console))

@dotnet/roslyn-infrastructure FYI I'll re-run.

```
14:31:56 xUnit output log: D:\j\workspace\windows_relea---c2abcdff\Binaries\Release\UnitTests\EditorServicesTest\xUnitResults\Roslyn.Services.Editor.UnitTests.dll.out.log
14:31:56 
xunit produced no error output but had exit code -1073740791
```

**Comment by jcouv** [comment]
test windows_release_unit32_prtest please

**Comment by heejaechang** [comment]
nice! it would be awesome if we enhance extract method to support "Extract Local Method" as well. basically letting users to extract method to local function rather than regular method.

**Comment by jcouv** [comment]
@heejaechang I agree. There is an issue tracking the "Extract Local Method" scenario: https://github.com/dotnet/roslyn/issues/18239

**Comment by jcouv** [comment]
@Pilchie @jaredpar for ask-mode approval for 15.7. Thanks
FYI the compiler change is to not display "private static" when displaying the parts for local function symbols.

**Comment by Pilchie** [comment]
Approved for 15.7 Preview 3 pending passing Jenkins run.

**Comment by jcouv** [comment]
test windows_coreclr_test_prtest  please

**Comment by jcouv** [comment]
@dotnet/roslyn-infrastructure `windows_coreclr_test_prtest` has been showing "Triggered. (71/80 on win2016-base)" for multiple hours. Any tips?

**Comment by jasonmalinowski** [comment]
@jcouv That test was removed by #25451; I guess Jenkins must have started to report status and once that got merged realized it could stop and left you in limbo. Safe to ignore.

**Pull request merged**